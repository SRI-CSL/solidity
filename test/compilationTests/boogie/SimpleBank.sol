pragma solidity ^0.4.23;

/**
 * @notice invariant __verifier_sum(user_balances) == this.balance
 */
contract SimpleBank {
    mapping(address=>uint) user_balances;

    function deposit() payable public {
        require(this != msg.sender);
        user_balances[msg.sender] += msg.value;
    }

    function withdraw_transfer() public {
        require(this != msg.sender);
        if (user_balances[msg.sender] > 0 && address(this).balance > user_balances[msg.sender]) {
            msg.sender.transfer(user_balances[msg.sender]);
            user_balances[msg.sender] = 0;
        }
    }

    function withdraw_call_incorrect() public {
        require(this != msg.sender);
        uint amount = user_balances[msg.sender];
        if (amount > 0 && address(this).balance > amount) {
            if (!msg.sender.call.value(amount)("")) {
                revert();
            }
            user_balances[msg.sender] = 0;
        }
    }

    function withdraw_call_correct() public {
        require(this != msg.sender);
        uint amount = user_balances[msg.sender];
        if (amount > 0 && address(this).balance > amount) {
            user_balances[msg.sender] = 0;
            if (!msg.sender.call.value(amount)("")) {
                revert();
            }
        }
    }
}
