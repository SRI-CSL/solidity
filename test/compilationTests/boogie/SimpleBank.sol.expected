Verifying SimpleBank::deposit: OK
Verifying SimpleBank::withdraw_transfer: OK
Verifying SimpleBank::withdraw_call_incorrect: ERROR
 - Source SimpleBank.sol, line 26, col 18: Invariant '__verifier_sum_uint256(user_balances) == this.balance' might not hold before external call.
Verifying SimpleBank::withdraw_call_correct: OK
Verifying SimpleBank::[implicit_constructor]: OK
